File: tests/t20-exa-vector.c
#include "exa.h"
#include "exa-memory.h"

#include <math.h>

#define M 10

int main(int argc,char *argv[])
{
  MPI_Init(NULL,NULL);

  if(argc!=2){
    printf("Usage: %s <backend>\n",argv[0]);
    exit(0);
  }

  exaHandle h;
  exaInit(&h,MPI_COMM_WORLD,argv[1]);

  exaVector vec;
  exaVectorCreate(h,M,exaScalar_t,&vec);

  if(M!=exaVectorGetSize(vec))
    fprintf(stderr,"exaVectorGetSize failed: %d != %d\n",M,
      exaVectorGetSize(vec));

  exaScalar *in,*out;
  exaCalloc(M,&in); exaCalloc(M,&out);

  int i;
  for(i=0;i<M;i++)
    in[i]=i+0.1;
  exaVectorWrite(vec,in);

  exaVectorRead(vec,out);
  for(i=0;i<M;i++)
    if(fabs(i+0.1-out[i])>EXA_TOL)
      fprintf(stderr,"Error %lf != %lf\n",i+0.1,out[i]);

  exaFree(in);
  exaFree(out);
  exaDestroy(vec);

  exaVector iVec;
  exaVectorCreate(h,M,exaInt_t,&iVec);

  if(M!=exaVectorGetSize(iVec))
    fprintf(stderr,"exaVectorGetSize failed: %d != %d\n",M,
      exaVectorGetSize(iVec));

  exaInt *iIn,*iOut;
  exaCalloc(M,&iIn); exaCalloc(M,&iOut);

  for(i=0;i<M;i++)
    iIn[i]=2*i+1;
  exaVectorWrite(iVec,iIn);

  exaVectorRead(iVec,iOut);
  for(i=0;i<M;i++)
    if(iOut[i]!=2*i+1)
      fprintf(stderr,"Error %d != %d\n",2*i+1,iOut[i]);

  exaFree(iIn);
  exaFree(iOut);
  exaDestroy(iVec);

  exaFinalize(h);

  MPI_Finalize();

  return 0;
}

File: tests/t20-exa-vector.f
      program t10_exa_handle
      implicit none

      include 'mpif.h'
      include 'exaf.h'

      integer ierr,nargs,i

      integer handle

      integer vector
      integer length
      integer*8 in_offset,out_offset
      parameter(in_offset=0)

      integer M
      parameter(M=10)

      real*8 in_(M),out_(M)

      character*128 arg

      call mpi_init(ierr)

      nargs=command_argument_count()
      if(nargs.ne.1) then
        write(stderr,*) "Usage: <program> <backend>"
        stop 1
      endif
      call get_command_argument(1,arg)

      call exainit(trim(arg),mpi_comm_world,handle,ierr)
      call exavectorcreate(handle,M,exa_scalar,vector,ierr)

      call exavectorgetsize(length,vector,ierr)
      if(length .ne. M) then
        write(stderr,*) 'exavectorgetsize failed.'
      endif

      do i=1,M
        in_(i) =i+0.1
        out_(i)=0.0
      enddo
      call exavectorwrite(vector,in_,in_offset,ierr)

      call exavectorread(vector,out_,out_offset,ierr)
      do i=1,M
        if(abs(out_(i)-(i+0.1))) then
          write(stderr,*) 'Error reading vector.'
        endif
      enddo

      call exavectorfree(vector,ierr)
      call exafinalize(handle,ierr)

      call mpi_finalize(ierr)
      end
